syft/pkg/cataloger/golang/scan_binary.go

```diff
@@ -89,6 +116,277 @@ func getCryptoSettingsFromVersion(v version.Version) []string {
+	rdr, err0 := basePath.Open(fileName)
+	if err0 != nil {
+		log.Debugf("error opening file %s: %v", fileName, err)
+		return
+	}
+	defer internal.CloseAndLogError(rdr, fileName)
+	var bytes []byte
+	reader := file.NewLocationReadCloser(file.NewLocation(fileName), rdr)
+	bytes, _ = io.ReadAll(reader)
+	if len(bytes) != 0 {
+		res = string(bytes)
+	}
+	return
+}
+
+func findVersionsInVendor(basePath fs.FS) (map[string]string, error) {
+	res := make(map[string]string)
+	err := fs.WalkDir(basePath, ".", func(filePath string, _ fs.DirEntry, err error) error {
+		if err != nil {
+			return fmt.Errorf("error when walking down the dir to find modules.txt")
+		}
+
+		if strings.EqualFold(filePath, "modules.txt") {
+			var contents string
+			contents, err = fetchFileContents(basePath, filePath)
+			if err == nil && len(contents) != 0 {
+				lines := strings.Split(contents, "\n")
+				for _, line := range lines {
+					// the target line is like this
+					// # github.com/google/go-cmp v0.7.0
+					if strings.HasPrefix(line, "# ") {
+						tuple := strings.Split(line, " ")
+						name := tuple[1]
+						ver := tuple[2]
+						res[name] = ver
+					}
+				}
+			}
+			return nil
+		}
+		return nil
+	})
+	if len(res) == 0 {
+		return nil, fmt.Errorf("modules.txt doesn't exist in the Go Vendors,hence no versioned path found for it")
+	}
+	return res, err
+}
+
+func (c *goBinaryCataloger) getModulesInfoInCache(syms []elf.Symbol, goPath fs.FS) map[string]*debug.Module {
+	uniqueModules := make(map[string]*debug.Module)
+	for _, sym := range syms {
+		// sym.Info is a common type in Linux, the low 4 bits represents type and the high 4 indicates binding
+		// the type for 0x12 is FUNC(0x02) and the binding is GLOBAL(0x01)
+		// since we seek for external dependencies, only global type symbols are needed
+		if sym.Info != 0x12 {
+			continue
+		}
+		urlDir, nameWithoutVersion := trimmedAsURL(sym.Name)
+		urlDirInPath := uncapitalize(urlDir)
+		nameInPath := uncapitalize(nameWithoutVersion)
+		// the sym is not a mark of third-party function
+		if len(urlDir) == 0 {
+			continue
+		}
+		modPair := fmt.Sprintf("%s/%s", urlDir, nameWithoutVersion)
+		modPairInPath := fmt.Sprintf("%s/%s", urlDirInPath, nameInPath)
+		if _, exists := uniqueModules[modPair]; exists {
+			continue
+		}
+		_, err := goPath.Open(urlDirInPath)
+		if err != nil {
+			continue
+		}
+		dir, _ := fs.Sub(goPath, urlDirInPath)
+
+		Version, err := findVersionInCache(dir, nameInPath)
+		// No such module
+		if err != nil || len(Version) == 0 {
+			continue
+		}
+
+		cachedir, err2 := fs.Sub(goPath, "cache/download/"+modPairInPath+"/@v")
+		if err2 != nil {
+			continue
+		}
+		checksum, _ := getCachedChecksum(cachedir, modPair, Version)
+		uniqueModules[modPair] = &debug.Module{
+			Path:    fmt.Sprintf("%s/%s", urlDir, nameWithoutVersion),
+			Version: Version,
+			Sum:     checksum,
+			Replace: nil,
+		}
+	}
+	return uniqueModules
+}
+
+func uncapitalize(name string) (newName string) {
```

Comment:
There is another function already doing this we should reuse: https://github.com/anchore/syft/blob/f9d0fa81ab6c2232df7b6b2d4b966797483f2ba5/syft/pkg/cataloger/golang/licenses.go#L293